File: sdks/go/qnsi/webhooks.go
package qnsp

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"strings"
	"time"
)

// MaxWebhookSkew is the default replay-protection window for ParseWebhook.
const MaxWebhookSkew = 5 * time.Minute

// WebhookEvent is the typed envelope every QNSP webhook follows.
type WebhookEvent struct {
	EventType  string         `json:"event_type"`
	EventID    string         `json:"event_id"`
	OccurredAt string         `json:"occurred_at"`
	Payload    map[string]any `json:"payload"`
	Raw        map[string]any `json:"-"`
}

// VerifyWebhookSignature compares the HMAC-SHA-256 of `body` keyed by
// `secret` against the value in the X-QNSP-Signature header. The header
// must be of the form "sha256=<hex>".
func VerifyWebhookSignature(body []byte, signatureHeader, secret string) error {
	if !strings.HasPrefix(signatureHeader, "sha256=") {
		return &WebhookError{Reason: "signature header must start with 'sha256='"}
	}
	expectedHex := strings.TrimPrefix(signatureHeader, "sha256=")
	expected, err := hex.DecodeString(expectedHex)
	if err != nil {
		return &WebhookError{Reason: "signature is not valid hex"}
	}
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(body)
	actual := mac.Sum(nil)
	if !hmac.Equal(actual, expected) {
		return &WebhookError{Reason: "signature mismatch"}
	}
	return nil
}

// ParseWebhook verifies the HMAC, enforces replay protection against
// `now`, parses the JSON body, and returns a typed WebhookEvent.
//
// Pass MaxWebhookSkew (or your own value) for `maxSkew`. Pass time.Time{}
// for `now` to use time.Now().
func ParseWebhook(body []byte, signatureHeader, timestampHeader, secret string, maxSkew time.Duration, now time.Time) (*WebhookEvent, error) {
	if err := VerifyWebhookSignature(body, signatureHeader, secret); err != nil {
		return nil, err
	}

	if timestampHeader != "" {
		if maxSkew == 0 {
			maxSkew = MaxWebhookSkew
		}
		ts, err := time.Parse(time.RFC3339, timestampHeader)
		if err != nil {
			ts, err = time.Parse("2006-01-02T15:04:05Z", timestampHeader)
		}
		if err != nil {
			return nil, &WebhookError{Reason: "timestamp header is not RFC3339"}
		}
		ref := now
		if ref.IsZero() {
			ref = time.Now().UTC()
		}
		delta := ref.Sub(ts)
		if delta > maxSkew {
			return nil, &WebhookError{Reason: "timestamp is too old"}
		}
		if -delta > maxSkew {
			return nil, &WebhookError{Reason: "timestamp is in the future"}
		}
	}

	var raw map[string]any
	if err := json.Unmarshal(body, &raw); err != nil {
		return nil, &WebhookError{Reason: "body is not valid JSON"}
	}

	eventType, _ := raw["event_type"].(string)
	if eventType == "" {
		return nil, &WebhookError{Reason: "missing event_type"}
	}
	eventID, _ := raw["event_id"].(string)
	if eventID == "" {
		return nil, &WebhookError{Reason: "missing event_id"}
	}
	occurredAt, _ := raw["occurred_at"].(string)
	payload, _ := raw["payload"].(map[string]any)
	if payload == nil {
		payload = map[string]any{}
	}

	return &WebhookEvent{
		EventType:  eventType,
		EventID:    eventID,
		OccurredAt: occurredAt,
		Payload:    payload,
		Raw:        raw,
	}, nil
}

File: sdks/rust/qnsi/src/webhooks.rs
//! QNSI webhook signature verification + typed event parsing.
//!
//! Every QNSI webhook is signed with HMAC-SHA-256 over the raw request
//! body. Always verify the **raw bytes** before parsing JSON.

use chrono::{DateTime, Duration, Utc};
use hmac::{Hmac, Mac};
use sha2::Sha256;

use crate::errors::WebhookError;

type HmacSha256 = Hmac<Sha256>;

/// Default replay-protection window for [`parse_webhook`].
pub const MAX_WEBHOOK_SKEW: Duration = Duration::minutes(5);

/// The typed envelope every QNSI webhook follows.
#[derive(Debug, Clone)]
pub struct WebhookEvent {
    pub event_type: String,
    pub event_id: String,
    pub occurred_at: String,
    pub payload: serde_json::Map<String, serde_json::Value>,
}

/// Constant-time HMAC-SHA-256 verification. The header must be of the
/// form `sha256=<hex>`.
pub fn verify_webhook_signature(
    body: &[u8],
    signature_header: &str,
    secret: &str,
) -> Result<(), WebhookError> {
    let expected_hex = signature_header.strip_prefix("sha256=").ok_or_else(|| {
        WebhookError {
            reason: "signature header must start with 'sha256='".into(),
        }
    })?;
    let expected = hex::decode(expected_hex).map_err(|_| WebhookError {
        reason: "signature is not valid hex".into(),
    })?;
    let mut mac = HmacSha256::new_from_slice(secret.as_bytes()).map_err(|_| WebhookError {
        reason: "invalid secret".into(),
    })?;
    mac.update(body);
    mac.verify_slice(&expected).map_err(|_| WebhookError {
        reason: "signature mismatch".into(),
    })?;
    Ok(())
}

/// Verify the HMAC, enforce replay protection, parse the JSON body, and
/// return a typed [`WebhookEvent`].
pub fn parse_webhook(
    body: &[u8],
    signature_header: &str,
    timestamp_header: Option<&str>,
    secret: &str,
    max_skew: Duration,
    now: Option<DateTime<Utc>>,
) -> Result<WebhookEvent, WebhookError> {
    verify_webhook_signature(body, signature_header, secret)?;

    if let Some(ts) = timestamp_header {
        let parsed = DateTime::parse_from_rfc3339(ts)
            .map_err(|_| WebhookError {
                reason: "timestamp header is not RFC3339".into(),
            })?
            .with_timezone(&Utc);
        let reference = now.unwrap_or_else(Utc::now);
        let delta = reference - parsed;
        if delta > max_skew {
            return Err(WebhookError {
                reason: "timestamp is too old".into(),
            });
        }
        if -delta > max_skew {
            return Err(WebhookError {
                reason: "timestamp is in the future".into(),
            });
        }
    }

    let raw: serde_json::Value = serde_json::from_slice(body).map_err(|_| WebhookError {
        reason: "body is not valid JSON".into(),
    })?;
    let obj = raw.as_object().ok_or_else(|| WebhookError {
        reason: "body is not a JSON object".into(),
    })?;

    let event_type = obj
        .get("event_type")
        .and_then(|v| v.as_str())
        .ok_or_else(|| WebhookError {
            reason: "missing event_type".into(),
        })?
        .to_string();
    let event_id = obj
        .get("event_id")
        .and_then(|v| v.as_str())
        .ok_or_else(|| WebhookError {
            reason: "missing event_id".into(),
        })?
        .to_string();
    let occurred_at = obj
        .get("occurred_at")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    let payload = obj
        .get("payload")
        .and_then(|v| v.as_object())
        .cloned()
        .unwrap_or_default();

    Ok(WebhookEvent {
        event_type,
        event_id,
        occurred_at,
        payload,
    })
}
